doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1506,160 +1494,140 @@ Managing the local context
 
-.. tacv:: pose proof @term {? as @simple_intropattern}
-   :name: pose proof
+      Arises when the argument :token:`type` is neither of type :g:`Prop`,
+      :g:`Set` nor :g:`Type`.
 
-   This tactic behaves like :n:`assert @type {? as @simple_intropattern} by exact @term`
-   where :token:`type` is the type of :token:`term`. In particular,
-   :n:`pose proof @term as @ident` behaves as :n:`assert (@ident := @term)`
-   and :n:`pose proof @term as @simple_intropattern` is the same as applying the
-   :token:`simple_intropattern` to :token:`term`.
+   .. exn:: Proof is not complete.
+      :name: Proof is not complete. (assert)
 
-.. tacv:: epose proof @term {? as @simple_intropattern}
-   :name: epose proof
+      :n:`@ltac_expr3` was not able to prove the new hypothesis.
 
-   While :tacn:`pose proof` expects that no existential variables are generated by
-   the tactic, :tacn:`epose proof` removes this constraint.
+   .. tacn:: eassert ( @ident : @type ) {? by @ltac_expr3 }
+             eassert ( @ident := @term )
+             eassert @one_term {? @as_ipat } {? by @ltac_expr3 }
+      :name: eassert; _; _
 
-.. tacv:: pose proof (@ident := @term)
+      Unlike :tacn:`assert`, the :n:`@type`\s, :n:`@term`\s and :n:`@one_term`\s in
+      :tacn:`eassert` may contain :term:`holes <hole>`, denoted by :n:`_`,
+      for which the tactic will create existential variables.  This lets you
+      avoid specifying the asserted statement completely before starting to
+      prove it.
 
-   This is an alternative syntax for :n:`assert (@ident := @term)` and
-   :n:`pose proof @term as @ident`, following the model of :n:`pose
-   (@ident := @term)` but dropping the value of :token:`ident`.
+.. tacn:: enough ( @ident : @type ) {? by @ltac_expr3 }
+          enough @one_term {? @as_ipat } {? by @ltac_expr3 }
+   :name: enough; _
 
-.. tacv:: epose proof (@ident := @term)
+   .. todo PR The only difference vs assert is where the goal is added?  If so, then
+      this entry should only refer to assert and mention the difference(s).
+      There are several similar tactics in this section.  Maybe we should add a
+      summary of their differences?  Perhaps you could briefly outline them?
 
-   This is an alternative syntax for :n:`eassert (@ident := @term)`
-   and :n:`epose proof @term as @ident`, following the model of
-   :n:`epose (@ident := @term)` but dropping the value of
-   :token:`ident`.
+   The first form adds a new hypothesis :n:`@ident : @type` to the context
+   and adds :n:`@type` as a new subgoal after the current goal (rather than
+   before it as :tacn:`assert` would).  If present, :n:`@ltac_expr3` must solve the
+   initial goal entirely after the extra assumption :n:`@type` is added and possibly
+   destructed; otherwise, the tactic will fail.
 
-.. tacv:: enough (@ident : @type)
-   :name: enough
+   In the second form, if :n:`@as_ipat` isn't specified, the tactic adds a new
```

Comment:
I'm generally OK with this description, except that now that the first paragraph only talks about the first form, this means that the second form is under-specified. There is a lot of content from the first paragraph that still applies to the second form, but it is the reader's job to guess.

Reply:
How about this?  I chose to describe `@ltac_expr3` in two places rather than in the first paragraph because `@as_ipat` only applies to the second form.

```
   Adds a new hypothesis to the current subgoal and a new subgoal to prove the
   hypothesis.  The new subgoal appears immediately after the current goal in
   the list of subgoals.

   The first form adds a new hypothesis :n:`@ident : @type`
   and :n:`@type` as the new subgoal.  Then, if :n:`@ltac_expr3` is
   specified, it applies that tactic to fully prove the current subgoal
   (and otherwise fails).

   In the second form, if :n:`@as_ipat` isn't specified, the tactic adds a new
   hypothesis :n:`@one_term` with a name chosen by Coq.  Otherwise, it transforms
   :n:`@one_term` as specified by :n:`@as_ipat` and adds the resulting new hypotheses.
   The :n:`@as_ipat` may also expand the current subgoal into multiple subgoals.
   Then, if :n:`@ltac_expr3` is specified, it is applied to and must succeed on all
   of them.
```

Reply:
LGTM. I would just not be that explicit about "immediately after the current goal in the list of subgoals" because a tactic can only create subgoals in the location of the goal it was applied on, so there would be no way that this was inserted later in the list of subgoals. But that's OK if you prefer to keep it.

Reply:
OK, dropped the "immediately" sentence.